src/ucioption.h

```diff
@@ -72,10 +56,38 @@ class Option {
     friend std::ostream& operator<<(std::ostream&, const OptionsMap&);
 
    private:
+    friend class OptionsMap;
+    friend class Engine;
+    friend class Tune;
+
+    void listen(OnUpdate&&);
+    void operator<<(const Option&);
+
     std::string defaultValue, currentValue, type;
     int         min, max;
     size_t      idx;
     OnChange    on_change;
+    OnUpdate    listener = nullptr;
+};
+
+class OptionsMap {
```

Comment:
We may want to strive to make Option and OptionsMap unrelated to UCI at all, and then they could be placed in a different header. We're pretty close.